File: app/src/lib/api/mock-sounds.ts
// The mock's sound catalog (#117): what `src/session/sounds.rs` does, built from the mock's
// fonts, plugins and sound library. No audio. MockSession calls `cmd` for its commands
// (running what it hands back), `advance` for the audition, and `derive` after every change.
// The twin of app/src-tauri/src/mock_sounds.rs.

import { guessCategory, presetsOf } from './mock-sound-library'
import { MAX_RECENTS, parsePresetId, pluginCategory, presetId, type SoundCatalog, type SoundEntry, type SoundsCmd, type SoundsState } from './sounds'
import type { PatchCategory } from './sound-library'
import type { AppCmd, AppState } from './types'

export function initialSounds(): SoundsState {
  return { revision: 0, count: 0, scanning: false, auditioning: null }
}

export class MockSounds {
  private favourites = new Set<string>()
  private recents: string[] = []
  private categories = new Map<string, PatchCategory>()
  private audition: { id: string; left: number } | null = null
  private key = ''
  private revision = 0

  catalog(st: AppState): SoundCatalog {
    const recent = (id: string) => this.recents.includes(id)
    const entries: SoundEntry[] = []
    for (const f of st.io.soundFonts) {
      for (const p of presetsOf(f)) {
        const id = presetId(f, p.bank, p.program)
        entries.push({ id, name: p.name, category: guessCategory(p.bank, p.program), source: 'soundFont', detail: f, favourite: this.favourites.has(id), recent: recent(id), plugin: null })
      }
    }
    for (const p of st.plugins.list) {
      const id = `au:${p.id}`
      entries.push({
        id,
        name: p.name,
        category: this.categories.get(id) ?? pluginCategory(p.name, p.manufacturer),
        source: 'plugin',
        detail: p.manufacturer,
        favourite: this.favourites.has(id),
        recent: recent(id),
        plugin: { format: p.format, lastError: p.lastError },
      })
    }
    for (const p of st.soundLibrary.patches) {
      const id = `saved:${p.id}`
      const detail = p.source.kind === 'soundFont' ? p.source.file : p.source.componentId
      entries.push({ id, name: p.name, category: p.category, source: 'saved', detail, favourite: p.favourite, recent: recent(id), plugin: null })
    }
    return { revision: this.revision, entries, recents: [...this.recents] }
  }

  private known(st: AppState, id: string): boolean {
    const pre = parsePresetId(id)
    if (pre) return st.io.soundFonts.includes(pre.file) && presetsOf(pre.file).some((p) => p.bank === pre.bank && p.program === pre.program)
    if (id.startsWith('au:')) return st.plugins.list.some((p) => `au:${p.id}` === id)
    if (id.startsWith('saved:')) return st.soundLibrary.patches.some((p) => `saved:${p.id}` === id)
    return false
  }

  /** Run a command: an error's text, or the commands the rest of the mock runs for it (an
   * `assignLastAdded` part: then give that part the patch just added). */
  cmd(st: AppState, c: SoundsCmd): { error?: string; run?: AppCmd[]; assignLastAdded?: number } {
    if (c.type === 'stopSoundAudition') {
      this.audition = null
      return {}
    }
    if (c.type === 'assignSound' && (c.part < 0 || c.part > 3)) return { error: `no keyboard part ${c.part} (0-3)` }
    if (!this.known(st, c.id)) return { error: `no sound ${c.id}` }
    const saved = c.id.startsWith('saved:') ? c.id.slice(6) : null
    switch (c.type) {
      case 'setSoundFavourite':
        if (saved) return { run: [{ type: 'setPatchFavourite', id: saved, favourite: c.on }] }
        if (c.on) this.favourites.add(c.id)
        else this.favourites.delete(c.id)
        return {}
      case 'auditionSound':
        if (st.transport.running) return { error: 'Stop the band to audition a sound' }
        this.audition = { id: c.id, left: 3000 }
        return {}
      case 'setSoundCategory': {
        if (saved) {
          const p = st.soundLibrary.patches.find((q) => q.id === saved)!
          return { run: [{ type: 'updatePatch', id: saved, patch: { name: p.name, category: c.category, tags: p.tags, favourite: p.favourite, source: p.source, defaults: p.defaults } }] }
        }
        if (!c.id.startsWith('au:')) return { error: "a preset's category is its GM family" }
        this.categories.set(c.id, c.category)
        return {}
      }
      case 'assignSound': {
        this.recents = [c.id, ...this.recents.filter((r) => r !== c.id)].slice(0, MAX_RECENTS)
        if (saved) return { run: [{ type: 'setPartPatch', part: c.part, id: saved }] }
        if (c.id.startsWith('au:')) return { run: [{ type: 'setPartPlugin', part: c.part, id: c.id.slice(3), state: null }] }
        const pre = parsePresetId(c.id)!
        if (st.io.soundFontFile === pre.file && pre.bank === 0) return { run: [{ type: 'setPartVoice', part: c.part, program: pre.program }] }
        const have = st.soundLibrary.patches.find((p) => p.source.kind === 'soundFont' && p.source.file === pre.file && p.source.bank === pre.bank && p.source.program === pre.program)
        if (have) return { run: [{ type: 'setPartPatch', part: c.part, id: have.id }] }
        return { run: [{ type: 'addPresetAsPatch', file: pre.file, bank: pre.bank, program: pre.program, name: null }], assignLastAdded: c.part }
      }
    }
  }

  /** The library patch a program map rule gets for catalog entry `id` (#117): a saved
   * sound's own, else the library's patch for the preset or plugin, added once through
   * `run`. An id without a catalog prefix is a patch id already. */
  patchFor(st: AppState, id: string, run: (c: AppCmd) => void): { patch: string } | { error: string } {
    if (id.startsWith('saved:')) return { patch: id.slice(6) }
    const pre = parsePresetId(id)
    if (!pre && !id.startsWith('au:')) return { patch: id }
    if (!this.known(st, id)) return { error: `no sound ${id}` }
    const plugin = pre ? null : id.slice(3)
    const have = st.soundLibrary.patches.find((p) =>
      pre ? p.source.kind === 'soundFont' && p.source.file === pre.file && p.source.bank === pre.bank && p.source.program === pre.program : p.source.kind === 'plugin' && p.source.componentId === plugin && !p.source.state,
    )
    if (have) return { patch: have.id }
    if (pre) run({ type: 'addPresetAsPatch', file: pre.file, bank: pre.bank, program: pre.program, name: null })
    else {
      const e = st.plugins.list.find((p) => p.id === plugin)!
      const category = this.categories.get(id) ?? pluginCategory(e.name, e.manufacturer)
      run({ type: 'createPatch', patch: { name: e.name, category, tags: [], favourite: false, source: { kind: 'plugin', componentId: e.id, state: '' }, defaults: { volume: null, pan: null, reverb: null, chorus: null, octave: 0 } } })
    }
    return { patch: st.soundLibrary.lastAdded ?? '' }
  }

  advance(ms: number, running: boolean) {
    if (!this.audition) return
    this.audition.left -= ms
    if (this.audition.left <= 0 || running) this.audition = null
  }

  /** `state.sounds`: a new revision whenever what the catalog is built from changed. */
  derive(st: AppState) {
    const key = JSON.stringify([st.io.soundFonts, st.io.soundFontFile, st.plugins.list, st.soundLibrary.patches, [...this.favourites], this.recents, [...this.categories]])
    if (key !== this.key || this.revision === 0) {
      this.key = key
      this.revision++
    }
    const presets = st.io.soundFonts.reduce((n, f) => n + presetsOf(f).length, 0)
    st.sounds = {
      revision: this.revision,
      count: presets + st.plugins.list.length + st.soundLibrary.patches.length,
      scanning: st.plugins.scanning,
      auditioning: this.audition?.id ?? null,
    }
  }
}

File: app/src-tauri/src/mock_sounds.rs
//! The mock's sound catalog (#117): what `src/session/sounds.rs` does, built from the
//! mock's fonts, plugins and sound library with the engine's own rules
//! (`yahaha::api::SoundPrefs`). No audio. The twin of `app/src/lib/api/mock-sounds.ts`.

use yahaha::api::*;
use yahaha::patches::PatchSource;

use super::sound::presets;

/// The catalog's settings, its revision and the audition.
#[derive(Default)]
pub struct MockSounds {
    prefs: SoundPrefs,
    audition: Option<(String, f64)>,
    /// What the catalog was last built from, and its revision.
    key: String,
    revision: u64,
}

/// What a `SoundsCmd` asks of the rest of the mock.
pub enum Then {
    Nothing,
    Run(Vec<AppCmd>),
    /// Add a preset to the library, then give keyboard part `.1` the patch added.
    AddThenAssign(AppCmd, u8),
}

impl MockSounds {
    fn fonts(st: &AppState) -> Vec<(String, Vec<Preset>)> {
        st.io.sound_fonts.iter().map(|f| (f.clone(), presets(f))).collect()
    }

    pub fn catalog(&self, st: &AppState) -> SoundCatalog {
        let fonts = Self::fonts(st);
        let fonts: Vec<(&str, &[Preset])> = fonts.iter().map(|(f, p)| (f.as_str(), p.as_slice())).collect();
        let patches: Vec<_> = st.sound_library.patches.iter().map(|p| p.patch.clone()).collect();
        SoundCatalog { revision: self.revision, entries: self.prefs.entries(&fonts, &st.plugins.list, &patches), recents: self.prefs.recents.clone() }
    }

    fn known(&self, st: &AppState, id: &str) -> bool {
        if let Some((file, bank, program)) = parse_preset_id(id) {
            st.io.sound_fonts.iter().any(|f| f == file) && presets(file).iter().any(|p| p.bank == bank && p.program == program)
        } else if let Some(plugin) = id.strip_prefix("au:") {
            st.plugins.list.iter().any(|p| p.id == plugin)
        } else if let Some(patch) = id.strip_prefix("saved:") {
            st.sound_library.patches.iter().any(|p| p.patch.id == patch)
        } else {
            false
        }
    }

    pub fn cmd(&mut self, st: &AppState, c: SoundsCmd) -> Result<Then, String> {
        let no = |id: &str| format!("no sound {id}");
        match c {
            SoundsCmd::SetSoundFavourite { id, on } => {
                if !self.known(st, &id) {
                    return Err(no(&id));
                }
                if let Some(patch) = id.strip_prefix("saved:") {
                    return Ok(Then::Run(vec![SoundLibraryCmd::SetPatchFavourite { id: patch.into(), favourite: on }.into()]));
                }
                if on {
                    self.prefs.favourites.insert(id);
                } else {
                    self.prefs.favourites.remove(&id);
                }
            }
            SoundsCmd::AuditionSound { id } => {
                if !self.known(st, &id) {
                    return Err(no(&id));
                }
                if st.transport.running {
                    return Err("Stop the band to audition a sound".into());
                }
                self.audition = Some((id, 3000.0));
            }
            SoundsCmd::StopSoundAudition => self.audition = None,
            SoundsCmd::AssignSound { part, id } => {
                if part > 3 {
                    return Err(format!("no keyboard part {part} (0-3)"));
                }
                if !self.known(st, &id) {
                    return Err(no(&id));
                }
                let then = if let Some(patch) = id.strip_prefix("saved:") {
                    Then::Run(vec![SoundLibraryCmd::SetPartPatch { part, id: Some(patch.into()) }.into()])
                } else if let Some(plugin) = id.strip_prefix("au:") {
                    Then::Run(vec![PluginCmd::SetPartPlugin { part, id: plugin.into(), state: None }.into()])
                } else {
                    let (file, bank, program) = parse_preset_id(&id).ok_or_else(|| no(&id))?;
                    if st.io.sound_font_file.as_deref() == Some(file) && bank == 0 {
                        Then::Run(vec![PartsCmd::SetPartVoice { part, program }.into()])
                    } else {
                        let existing = st.sound_library.patches.iter().find(|p| {
                            matches!(&p.patch.source, PatchSource::SoundFont { file: f, bank: b, program: q } if f == file && *b == bank && *q == program)
                        });
                        match existing {
                            Some(p) => Then::Run(vec![SoundLibraryCmd::SetPartPatch { part, id: Some(p.patch.id.clone()) }.into()]),
                            None => Then::AddThenAssign(SoundLibraryCmd::AddPresetAsPatch { file: file.into(), bank, program, name: None }.into(), part),
                        }
                    }
                };
                self.prefs.push_recent(id);
                return Ok(then);
            }
            SoundsCmd::SetSoundCategory { id, category } => {
                if !self.known(st, &id) {
                    return Err(no(&id));
                }
                if let Some(patch) = id.strip_prefix("saved:") {
                    let p = st.sound_library.patches.iter().find(|p| p.patch.id == patch).map(|p| p.patch.clone()).ok_or_else(|| no(&id))?;
                    let fields = PatchFields { name: p.name, category, tags: p.tags, favourite: p.favourite, source: p.source, defaults: p.defaults };
                    return Ok(Then::Run(vec![SoundLibraryCmd::UpdatePatch { id: p.id, patch: fields }.into()]));
                }
                if !id.starts_with("au:") {
                    return Err("a preset's category is its GM family".into());
                }
                self.prefs.sound_categories.insert(id, category);
            }
        }
        Ok(Then::Nothing)
    }

    /// The library patch a program map rule gets for catalog entry `id` (#117): a saved
    /// sound's own or the library's patch for the preset or plugin (`Ok(Ok(id))`), else
    /// the command that adds it (`Ok(Err(cmd))`; the patch is then `lastAdded`). An id
    /// without a catalog prefix is a patch id already.
    pub fn patch_for(&self, st: &AppState, id: &str) -> Result<Result<String, AppCmd>, String> {
        if let Some(patch) = id.strip_prefix("saved:") {
            return Ok(Ok(patch.into()));
        }
        let source = if let Some((file, bank, program)) = parse_preset_id(id) {
            PatchSource::SoundFont { file: file.into(), bank, program }
        } else if let Some(plugin) = id.strip_prefix("au:") {
            PatchSource::Plugin { component_id: plugin.into(), state: String::new() }
        } else {
            return Ok(Ok(id.into()));
        };
        if !self.known(st, id) {
            return Err(format!("no sound {id}"));
        }
        if let Some(p) = st.sound_library.patches.iter().find(|p| p.patch.source == source) {
            return Ok(Ok(p.patch.id.clone()));
        }
        Ok(Err(match source {
            PatchSource::SoundFont { file, bank, program } => SoundLibraryCmd::AddPresetAsPatch { file, bank, program, name: None }.into(),
            source @ PatchSource::Plugin { .. } => {
                let e = st.plugins.list.iter().find(|p| format!("au:{}", p.id) == id).ok_or_else(|| format!("no sound {id}"))?;
                let category = self.prefs.plugin_category(id, &e.name, &e.manufacturer);
                let patch = PatchFields { name: e.name.clone(), category, tags: vec![], favourite: false, source, defaults: Default::default() };
                SoundLibraryCmd::CreatePatch { patch }.into()
            }
        }))
    }

    pub fn advance(&mut self, ms: f64, running: bool) {
        if let Some((_, left)) = self.audition.as_mut() {
            *left -= ms;
            if *left <= 0.0 || running {
                self.audition = None;
            }
        }
    }

    /// `state.sounds`: a new revision whenever what the catalog is built from changed.
    pub fn derive(&mut self, st: &mut AppState) {
        let patches: Vec<_> = st.sound_library.patches.iter().map(|p| &p.patch).collect();
        let key = serde_json::to_string(&(&st.io.sound_fonts, &st.io.sound_font_file, &st.plugins.list, patches, &self.prefs)).unwrap_or_default();
        if key != self.key || self.revision == 0 {
            self.key = key;
            self.revision += 1;
        }
        let presets: usize = st.io.sound_fonts.iter().map(|f| presets(f).len()).sum();
        st.sounds = SoundsState {
            revision: self.revision,
            count: (presets + st.plugins.list.len() + st.sound_library.patches.len()) as u32,
            scanning: st.plugins.scanning,
            auditioning: self.audition.as_ref().map(|(id, _)| id.clone()),
        };
    }
}
